rgw/multisite: bucket instance cleanup after deletion

smanjara opened the pull request:
https://tracker.ceph.com/issues/20802





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

smanjara commented:
@cbodley this isn't ready yet. but I would like to get your feedback on the layout stuff.

smanjara commented on src/rgw/rgw_bucket_layout.h:
```diff
@@ -170,6 +173,10 @@ struct bucket_log_layout {
 
   bucket_index_log_layout in_index;
 
+  BucketLogType type2 = BucketLogType::Deleted;
```
@cbodley should we have a separate struct for the new type? I'm not even using type2 but I'm required to have the bucket_index_log_layout for the new type here to support encoding/decoding the BucketLogType

smanjara replied:
yup makes sense. thanks. done!

cbodley commented on src/rgw/driver/rados/rgw_trim_bilog.cc:
```diff
@@ -759,14 +760,25 @@ int BucketTrimInstanceCR::operate(const DoutPrefixProvider *dpp)
+
+  //remove bucket instance metadata
+  if (clean_info->first.layout.logs.front().layout.type == rgw::BucketLogType::Deleted) {
+  retcode = store->ctl()->bucket->remove_bucket_instance_info(bucket, clean_info->first, null_yield, dpp);
```
since this is in the `BucketTrimInstanceCR` coroutine, we can't call blocking functions here. `RGWGetBucketInstanceInfoCR` and `RGWPutBucketInstanceInfoCR` are used above to read/write this metadata - it looks like we need a corresponding `RGWRemoveBucketInstanceInfoCR`

smanjara replied:
done

smanjara commented on src/rgw/driver/rados/rgw_bucket.cc:
```diff
@@ -2021,6 +2021,31 @@ class RGWBucketMetadataHandler : public RGWBucketMetadataHandlerBase {
+
+    if (bi.deleted) {
+      bi.layout.logs.push_back(rgw::log_layout_from_deleted_index(0, bi.layout.current_index));
+    }
```
I can't seem to exercise the call path to do_remove() of RGWBucketInstanceMetadataHandler. I think I am missing something.

smanjara replied:
it seems like we don't call into `RGWBucketInstanceMetadataHandler::do_remove()` because the metadata_key we pass in through `RGWMetadataManager::remove() ` is  without an instance like metadata_key="bucket:some-bucket". whereas in metadata put() call we pass in the bucket instance.

cbodley replied:
right, bucket deletion in multisite only deletes the bucket entrypoint metadata, so that's the "bucket:some-bucket" key you're seeing

for the bucket instance metadata, the primary just adds the `BUCKET_DELETED` flag, so metadata sync would see that in `RGWBucketInstanceMetadataHandler::do_put()` -> `RGWMetadataHandlerPut_BucketInstance` instead

cbodley commented on src/rgw/rgw_common.h:
```diff
@@ -898,6 +898,7 @@ struct RGWBucketInfo {
   bool has_instance_obj{false};
   RGWObjVersionTracker objv_tracker; /* we don't need to serialize this, for runtime tracking */
   RGWQuotaInfo quota;
+  bool deleted{false};
```
we had discussed moving this `deleted` flag to an xattr to avoid changing the encoding, but it looks like it could be a flag in `RGWBucketFlags` for `RGWBucketInfo::flags`. that wouldn't change the encoding either

smanjara replied:
done!

cbodley commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -5135,6 +5135,19 @@ int RGWRados::delete_bucket(RGWBucketInfo& bucket_info, RGWObjVersionTracker& ob
 				       bucket_objs,
 				       cct->_conf->rgw_bucket_index_max_aio)();
+  } else {
+    bucket_info.flags |= BUCKET_DELETED;
+    map<string, bufferlist> attrs;
+
+    r = ctl.bucket->store_bucket_instance_info(bucket, bucket_info, y, dpp, RGWBucketCtl::BucketInstance::PutParams()
+                                                                    .set_exclusive(false)
+                                                                    .set_mtime(real_time())
+                                                                    .set_attrs(&attrs)
+                                                                    .set_orig_info(&bucket_info));
```
these metadata writes are protected by cls_version, so we may get `ECANCELED` if there are racing writes. we need a retry loop here that re-reads the info and adds back the `BUCKET_DELETED` flag. same goes for the `store_bucket_instance_info()` call that adds the 'deleted' log layout

`set_attrs()` with empty attrs will delete any existing attrs. if we don't have access to the real attrs, we'll need to modify callers to pass them in

smanjara replied:
got it

smanjara replied:
done

cbodley reviewed (commented):
the bucket layout and trim stuff looks great 👍

smanjara commented on src/rgw/driver/rados/rgw_bucket.cc:
```diff
@@ -2442,12 +2442,31 @@ class RGWBucketInstanceMetadataHandler : public RGWBucketInstanceMetadataHandler
+                                                 bci.info,
+                                                 nullptr,
+                                                 false,
+                                                 mtime, &bci.attrs, y, dpp);
```
pushed an update after validating that we do see the transition.

smanjara commented on src/rgw/rgw_bucket_layout.h:
```diff
@@ -208,6 +211,12 @@ inline auto matches_gen(uint64_t gen)
+inline bucket_log_layout_generation log_layout_from_deleted_index(
+    uint64_t gen, const bucket_index_layout_generation& index)
+{
+  return {gen, {BucketLogType::Deleted, {index.gen, index.layout.normal}}};
```
done

smanjara commented:
to-do: test the trim logic added and think about test cases.

cbodley commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -5130,6 +5130,33 @@ int RGWRados::delete_bucket(RGWBucketInfo& bucket_info, RGWObjVersionTracker& ob
 				       bucket_objs,
 				       cct->_conf->rgw_bucket_index_max_aio)();
+  } else {
+    bucket_info.flags |= BUCKET_DELETED;
+    static constexpr auto max_retries = 10;
+    int retries = 0;
+    do {
```
`bucket_info.flags |= BUCKET_DELETED;` needs to be inside the `do {` so it gets added back after `get_bucket_instance_info()`

cbodley replied:
p.s. it might be worth moving this stuff into a helper function so we don't over-complicate `RGWRados::delete_bucket()`

smanjara replied:
> `bucket_info.flags |= BUCKET_DELETED;` needs to be inside the `do {` so it gets added back after `get_bucket_instance_info()`

huh sorry we are overwriting, yes.

cbodley commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -5130,6 +5130,33 @@ int RGWRados::delete_bucket(RGWBucketInfo& bucket_info, RGWObjVersionTracker& ob
+    static constexpr auto max_retries = 10;
+    int retries = 0;
+    do {
+      r = ctl.bucket->store_bucket_instance_info(bucket, bucket_info, y, dpp, RGWBucketCtl::BucketInstance::PutParams()
+                                                                      .set_exclusive(false)
+                                                                      .set_mtime(real_time())
+                                                                      .set_attrs(&attrs)
+                                                                      .set_orig_info(&bucket_info));
+      if (r == -ECANCELED) {
+        //racing write. re-read bucket info
+        map<string, bufferlist> attrs;
+        int ret = get_bucket_instance_info(bucket, bucket_info, nullptr, &attrs, y, dpp);
```
shouldn't we reuse the existing `attrs` instead of reading and throwing away this local `attrs` variable?

smanjara replied:
yes, no reason to throw it away. done!

cbodley commented on src/rgw/driver/rados/rgw_sal_rados.cc:
```diff
@@ -330,8 +330,8 @@ int RadosBucket::remove(const DoutPrefixProvider* dpp,
   // if we deleted children above we will force delete, as any that
-  // remain is detritus from a prior bug
-  ret = store->getRados()->delete_bucket(info, ot, y, dpp, !delete_children);
+  // remain is detrius from a prior bug
```
detritus is the correct spelling. this was a conflict from https://github.com/ceph/ceph/pull/51229

smanjara replied:
bad rebase

cbodley commented on src/rgw/driver/rados/rgw_bucket.cc:
```diff
@@ -2597,15 +2605,37 @@ int RGWMetadataHandlerPut_BucketInstance::put_checked(const DoutPrefixProvider *
-                                                         pattrs,
-							 y,
-                                                         dpp);
+  //retry loop
```
sorry, i think i was mistaken about the need for a retry loop within `RGWMetadataHandlerPut_BucketInstance`

that should be happening at a higher level - for example, metadata sync already loops over calls to `put()` in [RGWMetaSyncSingleEntryCR](https://github.com/ceph/ceph/blob/785c108/src/rgw/driver/rados/rgw_sync.cc#L1314-L1317)

it doesn't make sense to loop here in `put_checked()`, because we're just rewriting the same thing we read without the other modifications from `put_check()` above. the retries from RGWMetaSyncSingleEntryCR should do the right thing already

smanjara replied:
done. I should have checked the sync call there. thanks!

cbodley commented:
looking great except for the s3select submodule change

cbodley commented on src/rgw/driver/rados/rgw_bucket.cc:
```diff
@@ -2560,6 +2563,26 @@ int RGWMetadataHandlerPut_BucketInstance::put_check(const DoutPrefixProvider *dp
+    if (bci.info.bucket_deleted() && log.layout.type != rgw::BucketLogType::Deleted) {
+      const auto index_log = bci.info.layout.logs.back();
+      const int shards_num = rgw::num_shards(index_log.layout.in_index);
+      bci.info.layout.logs.push_back({0, {rgw::BucketLogType::Deleted}});
```
the new log gen should probably be `log.gen + 1`

smanjara replied:
yup that was it. fixed!
I have created a testcase testing only secondary zone trimming, because we don't create the new layout on master zone. I think we will need to clean up instance by hand on the master zone or should we automate it in some way?

smanjara replied:
actually, manual clean up on master zone might be a bad idea, because if we forget to clean up, then during the next metadata sync, metadata master zone would end up syncing that bucket instance back to the secondary zones.

cbodley commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -5064,13 +5064,39 @@ int RGWRados::check_bucket_empty(const DoutPrefixProvider *dpp, RGWBucketInfo& b
   return 0;
 }
+
+int RGWRados::store_delete_bucket_info_flag(RGWBucketInfo& bucket_info, std::map<std::string, bufferlist>& attrs, optional_yield y, const DoutPrefixProvider *dpp) {
+  const rgw_bucket& bucket = bucket_info.bucket;
+  static constexpr auto max_retries = 10;
+  int retries = 0;
+  int r = 0;
+  do {
+    bucket_info.flags |= BUCKET_DELETED;
+    r = ctl.bucket->store_bucket_instance_info(bucket, bucket_info, y, dpp, RGWBucketCtl::BucketInstance::PutParams()
```
should this add a 'deleted' log layout too?

smanjara replied:
may be we could check if all the generations have been fully trimmed on the peer zones before calling metadata removal on master zone in the trim logic?

smanjara replied:
> may be we could check if all the generations have been fully trimmed on the peer zones before calling metadata removal on master zone in the trim logic?

huh no, if the bucket metadata doesn't exist, there will be nothing left to check. wonder what the StatusShards would return in that case.

smanjara replied:
> ok but if trim runs on the master zone first and clears the bucket metadata, that would end up syncing this metadata delete op even before we trim the bilog on the other zones.

this will only happen if we add the delete op to datalog entry.

smanjara replied:
> huh no, if the bucket metadata doesn't exist, there will be nothing left to check. StatusShards should return -ENOENT in that case.

we may need some special handling in this case, in bilog trim logic.

smanjara commented:
from my discussion with @cbodley, there are two pending issues to be addressed:
1. handle -ENOENT response from peer zones when the bucket instance metadata has been removed. for this, create a new coroutine wrapper that reads the StatusShards, if the response is ENOENT, mark the generation to a high number to skip that zone from further processing.  we would still want to be able to get the min marker for trimming from other valid zones.
2. handle metadata deletion on master zone. if the master has deleted the bucket instance, prevent it from being synced to other zones. other zones might still have their bilog trim pending and syncing bucket instance deletion from master zone, would end up deleting it on other zones as well.

smanjara commented:
> from my discussion with @cbodley, there are two pending issues to be addressed:
> 
> 1. handle -ENOENT response from peer zones
> 
> * [x]  when the bucket instance metadata has been removed. for this, create a new coroutine wrapper that reads the StatusShards, if the response is ENOENT, mark the generation to a high number to skip that zone from further processing.  we would still want to be able to get the min marker for trimming from other valid zones.
> 
> 2. handle metadata
> 
> * [x]  deletion on master zone. if the master has deleted the bucket instance, prevent it from being synced to other zones. other zones might still have their bilog trim pending and syncing bucket instance deletion from master zone, would end up deleting it on other zones as well.

@cbodley this is mostly complete. I have to resolve the failure coming from one of the testcases I have added. could you please review [eb2ee95](https://github.com/ceph/ceph/pull/53799/commits/eb2ee95a2fc37926ba1d923f7e1fe87f0d7bfaff) whenever you can?

cbodley commented on src/rgw/driver/rados/rgw_trim_bilog.cc:
```diff
@@ -495,9 +558,16 @@ class BucketTrimInstanceCR : public RGWCoroutine {
+      min_generation = min_generation;
+    }
+
+    ldpp_dout(dpp, 10) << "min_generation is " << min_generation << dendl;
```
i guess this part handles the case where all peers returned ENOENT so we get UINT64_MAX instead of a valid gen?

in that case, we know that all peers are caught up, so all of our log generations are eligible for removal

we need to make sure that `maybe_remove_generation()` picks the oldest log generation for cleanup. otherwise we'll skip the `if (clean_info) {` block below and enter the else-block that just looks for bilog entries to trim

if we leave `min_generation` at `logs.back().gen`, it seems like we'd clean all earlier generations, but never clean that last one to trigger bucket instance deletion. am i missing something else that makes this work?

i'm thinking this could just have a special case for UINT64_MAX that skips the std::find_if()
```c++
  if (min_generation == UINT64_MAX) {
    // if all peers have deleted this bucket, purge the rest of our log generations
    totrim.gen = UINT64_MAX;
    return 0;
  }
```

smanjara replied:
sorry for responding late. yes, that takes care of the ENOENT return case. thanks @cbodley
I don't see anything else pending here.

cbodley commented:
thanks @smanjara, the latest changes look great. i think we're missing something at the design level, though

where sync policies are in place, a bucket replication policy may only replicate a subset of the objects (or none at all) from zone A -> B. if zone B deletes its objects and then the bucket, there would still be extra objects present in zone A. as implemented, i think this pr would go ahead and delete all of the bucket metadata, leaving those objects orphaned on A

i don't remember whether we discussed this problem during the design phase, do you?

smanjara commented:
> thanks @smanjara, the latest changes look great. i think we're missing something at the design level, though
> 
> where sync policies are in place, a bucket replication policy may only replicate a subset of the objects (or none at all) from zone A -> B. if zone B deletes its objects and then the bucket, there would still be extra objects present in zone A. as implemented, i think this pr would go ahead and delete all of the bucket metadata, leaving those objects orphaned on A
> 
> i don't remember whether we discussed this problem during the design phase, do you?

@cbodley don't think we discussed this. if zone A still has the bucket and some objects in it, that means we also have the bucket entrypoint. before cleaning up bucket metadata, we could check if the local bucket entrypoint exists or not on zone A during trim?

cbodley commented:
> if zone A still has the bucket and some objects in it, that means we also have the bucket entrypoint

the problem is that when we send a DeleteBucket request to a zone B, only that zone checks to make sure it's empty. so even though zone A still has objects in it, zone B is allowed to successfully delete the bucket. that includes the removal of its bucket entrypoint

in the absence of sync policy, this is fine because zone B had to delete all of its objects to empty the bucket. we know that zone A will eventually sync those deletions even though we delete the bucket entrypoint

smanjara commented:
> 
> the problem is that when we send a DeleteBucket request to a zone B, only that zone checks to make sure it's empty. so even though zone A still has objects in it, zone B is allowed to successfully delete the bucket. that includes the removal of its bucket entrypoint
> 

@cbodley, with sync policy in place, i think we only have to care about our own bucket's entrypoint, which will still exist if there are objects in it, even if the peer zone's bucket is deleted. so, if bilog trim finds that we still have our local entrypoint i.e. `pbucket's` entrypoint, then we could skip the cleanup.

smanjara commented:
> > the problem is that when we send a DeleteBucket request to a zone B, only that zone checks to make sure it's empty. so even though zone A still has objects in it, zone B is allowed to successfully delete the bucket. that includes the removal of its bucket entrypoint
> 
> @cbodley, with sync policy in place, i think we only have to care about our own bucket's entrypoint, which will still exist if there are objects in it, even if the peer zone's bucket is deleted. so, if bilog trim finds that we still have our local entrypoint i.e. `pbucket's` entrypoint, then we could skip the cleanup.

got your point. do we want to skip over deleting bucket instance entirely if we find that the other zone's bucket is not empty? we could send a request to list remote bucket like how we do with RGWListRemoteBucketCR() in data_sync, during bilog trim, but only return a true or false for response.

smanjara commented:
> > @cbodley got your point. do we want to skip over deleting bucket instance entirely if we find that the other zone's bucket is not empty? we could send a request to list remote bucket like how we do with RGWListRemoteBucketCR() in data_sync, during bilog trim, but only return a true or false for response.
> 
> or to avoid listing, there is also RGWStatBucket_ObjStore_S3() api to get us the object count.

@cbodley is this reasonable? what do you think?

smanjara commented:
updated PR with remote bucket ENOTEMPTY check. working on tool to cleanup objects left behind.

cbodley commented on src/rgw/driver/rados/rgw_trim_bilog.cc:
```diff
@@ -1143,11 +1307,77 @@ int BucketTrimCR::operate(const DoutPrefixProvider *dpp)
       }
     }
 
+    // check for orphaned objects in buckets. if there are objects add the bucket to the omap list
+    // "deleted_buckets" object.
+
+    using ReadCR = RGWSimpleRadosReadCR<DeletedBucketList>;
```
BucketTrimCR runs on a fixed interval of ~20 minutes. i don't think we want to pay for all of this every time. as this omap list of deleted buckets grows, we'll have less time to spend on the actual trimming

i think it's best to leave this cleanup for the admin

smanjara replied:
agreed

smanjara replied:
@cbodley to prevent the omap object from growing too big, I was hoping that every time the command is run, we remove entries that don't have objects in them anymore . for that to happen, I tried listing those buckets in rgw_admin [L7397](https://github.com/ceph/ceph/pull/53799/commits/da328c2a2d08b636445398f8753bfd7ede735113#diff-f3b81fb31115c2616085efcfd6a8daed67a94d1041a03af5de48cc572978f547R7397). but we remove the bucket index objects as part of trimming in `BucketCleanIndexCollectCR()` . so any listing after that call would fail.

smanjara replied:
please review 92dc547 and da328c2a2

smanjara replied:
@cbodley could you please take a look at 461957aff008827c77621468c8b657132588d7b6 https://github.com/ceph/ceph/commit/92dc5474248d6c2f2cb0dcf696048f98d9e035d4 and https://github.com/ceph/ceph/commit/da328c2a2d08b636445398f8753bfd7ede735113

smanjara commented on src/rgw/driver/rados/rgw_trim_bilog.cc:
```diff
@@ -769,6 +855,15 @@ int BucketTrimInstanceCR::operate(const DoutPrefixProvider *dpp)
+        if (retcode < 0) {
+          ldpp_dout(dpp, 10) << "failed to list bucket: " << bucket_instance
+            << cpp_strerror(retcode) << dendl;
+          return set_cr_error(retcode);
```
i don't think these functions have to return error. they are not strictly necessary for trim to be successful.

cbodley reviewed (commented):
oops, i left some comments on 461957aff008827c77621468c8b657132588d7b6 but they aren't showing up in the review

smanjara commented on src/rgw/driver/rados/rgw_trim_bilog.cc:
```diff
@@ -759,14 +905,54 @@ int BucketTrimInstanceCR::operate(const DoutPrefixProvider *dpp)
+        return set_cr_error(retcode);
+        }
+    } else {
+      // update and store the object
+      deleted_buckets.entries.push_back(bucket_instance);
+      ldpp_dout(dpp, 10) << "added bucket:" << bucket_instance << " to deleted_buckets list: " << dendl;
+      using WriteCR = RGWSimpleRadosWriteCR<DeletedBucketList>;
+      yield call(new WriteCR(dpp, store,
+          rgw_raw_obj(store->svc()->zone->get_zone_params().log_pool, DeletedBucketList::oid),
+          deleted_buckets, &objv_tracker));
```
ok yeah I see the difference. thanks!

smanjara replied:
done

smanjara commented:
> oops, i left some comments on [461957a](https://github.com/ceph/ceph/commit/461957aff008827c77621468c8b657132588d7b6) but they aren't showing up in the review

@cbodley addressed all comments. please review.

cbodley commented on src/rgw/driver/rados/rgw_trim_bilog.cc:
```diff
@@ -1443,3 +1618,62 @@ int bilog_trim(const DoutPrefixProvider* p, rgw::sal::RadosStore* store,
+  }
+
+  librados::ObjectWriteOperation op;
+  for(const auto& key : buckets) {
+    op.omap_rm_keys(std::set<std::string>({key}));
+    r = rgw_rados_operate(dpp, ref.ioctx, ref.obj.oid, &op, y);
```
remove_deleted_buckets() is only removing these omap keys, but the bucket instance and index still exist, presumably with objects remaining. how do you see picture the admin cleaning those up? can we try to document this process?

smanjara replied:
right, this option `--remove-deleted` has to be used only once the admin has deleted the objects inside the bucket and after the trim has finished processing the bucket. but probably its hard for admin to say if the trim has finished processing the bucket. since we didn't want this to be a part of trimming process, I added it as a cli option. maybe sufficient documentation around this option is enough?

cbodley replied:
> maybe sufficient documentation around this option is enough?

could be. or maybe the `radosgw-admin bucket rm` command could try to remove the bucket from this omap object automatically

smanjara replied:
> > maybe sufficient documentation around this option is enough?
> 
> could be. or maybe the `radosgw-admin bucket rm` command could try to remove the bucket from this omap object automatically

the bucket omap entry is added only after trim sees `BucketLogType::Deleted` following a delete bucket execution. when do you suggest   `radosgw-admin bucket rm` has to be run to remove the entry? i could, instead, write up a separate command to handle its removal.

cbodley replied:
if we're documenting the use of `radosgw-admin bucket rm --bucket {name} --bucket-id {instance-id} --purge-objects` to clean these up, it would be nice if that automatically unlinked the bucket from this deleted_buckets object

smanjara replied:
what if an admin has used awscli or something else to delete the bucket? do we want them to run` bucket rm ` for the omap entry cleanup?

cbodley replied:
to calrify, i'm talking about deleted buckets that trim discovers to still contain objects. these get linked into `deleted_buckets`, so become visible to the `radosgw-admin bucket list-deleted` command

in response, the admin should check whether this bucket contains any important objects that need to be backed up elsewhere before completing the bucket's cleanup. once they're confident that they're recovered everything that was important, they can complete the cleanup with `radosgw-admin bucket rm --bucket {name} --bucket-id {instance-id} --purge-objects`. the bucket entrypoint was already deleted, so they have to specify the --bucket-id manually to find it. that will delete the remaining objects, then delete the bucket index object and instance metadata

after doing that for every bucket, they could run `bucket list-deleted --remove-deleted` to clear the list. but there's a risk that trim added another bucket to the list since they last ran `bucket list-deleted`, and `--remove-deleted` would clear that before they have a chance to do the cleanup

so i'm suggesting that the `bucket rm` command, on success, unlinks the bucket from this `deleted_objects` list to avoid these races and simplify the process for admins

smanjara replied:
oh ok I don't think we spoke about using `bucket rm ` earlier to clean up stuff locally. I was working with the assumption that once an admin moves/removes the objects inside the bucket, the next trim run will take care of the instance cleanup. this makes sense.

smanjara replied:
done. need to work on documentation.

smanjara replied:
updated PendingReleaseNotes. will run it through teuthology. do you have any more review comments @cbodley

cbodley commented on src/rgw/driver/rados/rgw_trim_bilog.cc:
```diff
@@ -759,14 +904,45 @@ int BucketTrimInstanceCR::operate(const DoutPrefixProvider *dpp)
+        }
+    } else {
+      yield {
+        // if there are objects present, add the bucket entry to omap list 'deleted_buckets'
+        auto& pool = store->svc()->zone->get_zone_params().log_pool;
+        const bufferlist in_bl;
+        std::map<std::string, bufferlist> bucket_entry{{bucket_instance, in_bl}};
+        ldpp_dout(dpp, 10) << "storing bucket in omap: " << bucket_instance << dendl;
+        call(new RGWRadosSetOmapKeysCR(store, rgw_raw_obj(pool, oid_prefix), bucket_entry));
```
does this `bucket_instance` string contain the bucket name, or just the bucket instance id?

to complete this cleanup, i think we'd expect the admin to run a command like:
```
$ radosgw-admin bucket rm --bucket {name} --bucket-id {instance-id} --purge-objects
```
which requires the bucket name too

smanjara replied:
it has tenant, bucket name and id. like this:

```
[
    "testx/bucket100:96c7bca1-54be-49e9-9cf2-ebbbd0153aea.4451.102",
    "testx/bucket102:96c7bca1-54be-49e9-9cf2-ebbbd0153aea.4451.104"
]
```

smanjara commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -5166,13 +5166,131 @@ int RGWRados::check_bucket_empty(const DoutPrefixProvider *dpp, RGWBucketInfo& b
+      rgw_obj_key obj;
+
+      if (rgw_obj_key::oid_to_key_in_ns(entries_iter->key.name, &obj, ns)) {
+        return -ENOTEMPTY;
```
@cbodley how about just parsing the oid instead? I have added that in commit 3d2b94f8da54e762b2a9d88e8bf268ce462e00ad do you see any problems with that?

cbodley replied:
sorry, i think we're confusing the two different layers of bucket listings. the ones we request directly from rados are unfiltered, so may contain hidden/namespaced entries (like the parts of incomplete multipart uploads that haven't turned into real objects yet). when we request listings from the S3 API, those hidden entries do get filtered out of the response

since we're fetching the S3 listings with RGWStatRemoteBucketCR, we shouldn't need to do any extra filtering here. we should just be able to check if `!list_result.entries.empty()` for the ENOTEMPTY case

smanjara replied:
@cbodley fixed. thanks! code is complete now, hopefully. will run it through teuthology and report soon.

smanjara commented:
tests for two zone configuration are passing:
http://qa-proxy.ceph.com/teuthology/smanjara-2024-12-05_16:51:12-rgw:multisite-wip-shilpa-cleanup-bi-distro-default-smithi/8021746/teuthology.log

multi zonegroup config fail the trim tests. looks like the test needs modification. 
http://qa-proxy.ceph.com/teuthology/smanjara-2024-12-05_16:51:12-rgw:multisite-wip-shilpa-cleanup-bi-distro-default-smithi/8021745/teuthology.log

but trimming on two zonegroup config works fine when tested by hand.

cbodley commented on src/test/rgw/rgw_multi/tests.py:
```diff
@@ -1664,6 +1673,118 @@ def make_test_bucket():
+    # delete bucket and test bilog autotrim
+    primary.conn.delete_bucket(test_bucket.name)
+
+    zonegroup_data_checkpoint(zonegroup_conns)
+    zonegroup_meta_checkpoint(zonegroup)
+
+    # run bilog trim twice to clean up previous gen
+    bilog_autotrim(secondary.zone)
+    time.sleep(config.checkpoint_delay)
+
+    # trim bilog on primary
+    bilog_autotrim(primary.zone)
+    time.sleep(config.checkpoint_delay)
+    
+    bilog_autotrim(secondary.zone)
```
i tend to think we should do the `bilog_autotrim()` and `check_bucket_instance_metadata()` on every zone, not just `rw_zones[0]` and `rw_zones[1]`. this would help make sure we get full test coverage for multi-zonegroups

clwluvw replied:
for multi-zonegroups shouldn't we just skip other zonegroups as they can't host any data for those buckets hosted in the bucket's zonegroup?

smanjara replied:
@clwluvw currently, the bucket would still create indexes on other zonegroups. so upon bucket deletion, we should also remove all the metadata associated with it.

clwluvw replied:
I don't think so, from here (https://github.com/ceph/ceph/blob/90d0f0306f9a1fe6e78bdda5c29fc5137dcafda6/src/rgw/rgw_op.cc#L3639-L3642) only when bucket's zonegroup is the RGW's zonegroup `createparams.zone_placement` will be assigned and then in `RGWRados::create_bucket` it only creates the indexes when `zone_placement` is not empty: https://github.com/ceph/ceph/blob/90d0f0306f9a1fe6e78bdda5c29fc5137dcafda6/src/rgw/driver/rados/rgw_rados.cc#L2362-L2367

clwluvw replied:
Do you have those commits In your branch?
It works for me as i have faced problems like this: https://github.com/ceph/ceph/pull/60227

clwluvw replied:
Can we have this commit (https://github.com/ceph/ceph/pull/59911/commits/be9ad44e5a7abb5f9d14ea89a6d46f4655c99c5e) or a check like this (https://github.com/ceph/ceph/blob/be9ad44e5a7abb5f9d14ea89a6d46f4655c99c5e/src/rgw/driver/rados/rgw_bucket.cc#L2820-L2824) backported here so we can make sure those indexes won't be created even through the meta sync?

smanjara replied:
as far as this pr is concerned, the cleanup works perfectly well on multiple zonegroups. so just making the test cases pass is what is needed here. even if the bucket is indexless, we will still have bucket instance left behind to be cleaned up. maybe we can consider indexless strategy separately outside this pr if it is well-motivated.

cbodley replied:
for multi-zonegroup, i could see a design where we let metadata sync (via `RGWBucketInstanceMetadataHandler::remove()`) delete the bucket instance metadata for buckets that belong to other zonegroups, and only defer that deletion until bilog trimming for local buckets. but i'm not sure it's worth treating local/remote buckets differently if the trimming solution can work for both

but who deletes the bucket index shards? i had thought `RGWBucketInstanceMetadataHandler::remove()` handled that, but i don't see a call to `clean_index()` there

clwluvw replied:
> but i'm not sure it's worth treating local/remote buckets differently if the trimming solution can work for both

I had a concern about unnecessary calls when it came to the scale of x zonegroups and x deletions. Perhaps hard to measure now but if we can eliminate why not? :)

> but who deletes the bucket index shards? i had thought RGWBucketInstanceMetadataHandler::remove() handled that, but i don't see a call to clean_index() there

We could add the if clause I mentioned earlier to prevent initializing indexes for remote buckets. I originally suggested using it for an indexless approach, but even if we don't take that route now, the check is still worth having. so we won't need to clean_index() at all for remote_buckets.

cbodley replied:
> Perhaps hard to measure now but if we can eliminate why not? :)

yeah.. bilog trimming makes its decisions based on the peer zones' sync status with respect to the local bilog. but when it runs on a remote zonegroup, there is no local bilog so no peers will report any status for it. empty status probably still causes us to handle the cleanup, but i agree that those http requests to fetch sync status aren't useful there

> so we won't need to clean_index() at all for remote buckets.

right, but we do need to call it for local buckets. i _think_ that should be part of the bucket instance metadata's deletion, which itself knows whether its indexless so can do the right thing

clwluvw replied:
> right, but we do need to call it for local buckets. i think that should be part of the bucket instance metadata's deletion, which itself knows whether its indexless so can do the right thing

Right, but I got confused why do we need to keep indexes when `is_syncing_bucket_meta()` and not call `CLSRGWIssueBucketIndexClean()`? I guess the status object is not stored in index so what is the point of having keeping it?

cbodley replied:
> right, but we do need to call it for local buckets. i _think_ that should be part of the bucket instance metadata's deletion, which itself knows whether its indexless so can do the right thing

sorry, Shilpa clarified this in the refactoring call. trim is already handling index object deletion via `BucketCleanIndexCollectCR`

clwluvw replied:
Just got a general question regarding `BucketCleanIndexCollectCR`, in multi-zonegroup (two zonegroup one in each) scenario in the current main branch implementation how is `BucketCleanIndexCollectCR` going to be called? Isn't that going to be stopped here: https://github.com/ceph/ceph/blob/7e6c5b8573955b5bdc5b6ca74aa5a06233a6f2cf/src/rgw/driver/rados/rgw_trim_bilog.cc#L619-L630

smanjara replied:
`RGWBucketGetSyncPolicyHandlerCR` is just reading the local zone's sync policy and returns a default handler if there isn't one. further down, we query peer zones of that zonegroup for sync status with a `get_all_dests()` and expect a response based on which we decide whether to trim a particular generation or not.
if you are asking about having a single zone in each zonegroup, RGWBucketGetSyncPolicyHandlerCR will still give you a default(empty) policy. so won't error out here. but it will go ahead and simply trim everything since it doesn't have any peers to query. this is something we will have to consider in cross-zonegroup replication.

clwluvw replied:
right, I forgot we create one policy handler per zone on startup https://github.com/ceph/ceph/blob/c0dae4614e22751b835928fee65d49b4a015f090/src/rgw/services/svc_zone.cc#L300

smanjara commented on src/rgw/driver/rados/rgw_bucket.cc:
```diff
@@ -2915,23 +2937,7 @@ int RGWBucketInstanceMetadataHandler::put_post(
-  std::ignore = update_bucket_topic_mappings(dpp, &bci, /*current_bci=*/nullptr,
-                                             driver);
-  return 0;
+  return 0; // skip bucket instance removal. each zone will handle it independently during trimming
```
yeah raw_key doesn't say much about the entity type.

cbodley commented on PendingReleaseNotes:
```diff
@@ -56,6 +56,10 @@
+* RGW: deleted buckets are automatically cleaned up as part of trimming process.
+  DeleteBucket will start returning 409 BucketNotEmpty errors until empty
+  on all zones when sync policy is enabled.
+  `radosgw-admin bucket list-deleted` lists such buckets.
```
we reverted this `bucket list-deleted` stuff, right?

smanjara replied:
yes I did.

cbodley commented on src/rgw/driver/rados/rgw_bucket.cc:
```diff
@@ -2869,6 +2871,26 @@ int RGWBucketInstanceMetadataHandler::put_prepare(
+      ldpp_dout(dpp, 10) << "store log layout type: " <<  bci.info.layout.logs.back().layout.type << dendl;
+      for (int i = 0; i < shards_num; ++i) {
+        ldpp_dout(dpp, 10) << "adding to data_log shard_id: " << i << " of gen:" << index_log.gen << dendl;
+        int ret = svc_datalog->add_entry(dpp, bci.info, index_log, i,
+                                                    null_yield);
```
this function has an `optional_yield y` that we can pass instead of `null_yield`

cbodley commented on src/rgw/driver/rados/rgw_bucket.cc:
```diff
@@ -2948,23 +2970,7 @@ int RGWBucketInstanceMetadataHandler::put_post(
-      entry, bci.info, &bci.info.objv_tracker, y, dpp);
-  if (ret < 0)
-    return ret;
-  std::ignore = update_bucket_topic_mappings(dpp, &bci, /*current_bci=*/nullptr,
-                                             driver);
```
`RGWBucketInstanceMetadataHandler::remove()` defers the bucket instance removal until trim, but do we ever make this call to `update_bucket_topic_mappings()` later?

maybe we should keep the `read_bucket_instance_info()` and `update_bucket_topic_mappings()` calls here, and just skip the `remove_bucket_instance_info()`?

smanjara replied:
yes this was a question I had. will add it back.

cbodley commented on src/rgw/driver/rados/rgw_cr_rados.h:
```diff
@@ -1648,3 +1708,105 @@ class RGWDataPostNotifyCR : public RGWCoroutine {
+    JSONDecoder::decode_json("Prefix", prefix, obj);
+    JSONDecoder::decode_json("MaxKeys", max_keys, obj);
+    JSONDecoder::decode_json("IsTruncated", is_truncated, obj);
+    JSONDecoder::decode_json("allow-unordered", allow_unordered, obj);
```
nit: `allow-unordered` is a request parameter, not part of the response

smanjara replied:
considering that we list only one object, may be I'd remove this parameter.

smanjara replied:
done

cbodley commented on src/test/rgw/rgw_multi/tests.py:
```diff
@@ -1665,6 +1676,128 @@ def make_test_bucket():
+
+    bilog_autotrim(primary.zone)
+
+    for zonegroup in realm.current_period.zonegroups:
+        zonegroup_conns = ZonegroupConns(zonegroup)
+        zonegroup_meta_checkpoint(zonegroup)
+
+        for zone in zonegroup_conns.zones:
+            log.info('trimming on zone=%s', zone.name)
+            bilog_autotrim(zone.zone)
+            time.sleep(config.checkpoint_delay)
+
+    # run bilog trim twice on primary zone where the bucket was resharded
+    bilog_autotrim(primary.zone)
+    
+    for zonegroup in realm.current_period.zonegroups:
+        for zone in zonegroup_conns.zones:
```
this is reusing the `zonegroup_conns` variable from the previous loop. i think we need to reinitialize it down here:
```diff
     for zonegroup in realm.current_period.zonegroups:
+        zonegroup_conns = ZonegroupConns(zonegroup)
         for zone in zonegroup_conns.zones:
```
same for the `test_bucket_log_trim_after_delete_bucket_secondary_reshard()` test below

smanjara replied:
oh yes

smanjara commented:
with https://github.com/ceph/ceph/pull/53799/commits/6544c69e9bf3a8f8af71bb9e33901bfc337fa07b the tests are passing:
https://pulpito.ceph.com/smanjara-2024-12-17_09:39:58-rgw:multisite-wip-shilpa-cleanup-bi-distro-default-smithi/

cbodley commented:
> with [6544c69](https://github.com/ceph/ceph/commit/6544c69e9bf3a8f8af71bb9e33901bfc337fa07b) the tests are passing: https://pulpito.ceph.com/smanjara-2024-12-17_09:39:58-rgw:multisite-wip-shilpa-cleanup-bi-distro-default-smithi/

i do love to see green multisite runs!

the test coverage for cleanup on `bilog autotrim` looks good, but we don't have any coverage for the new `BucketNotEmpty` check. i was hoping we could copy one of Soumya's sync policy test cases for this, but it looks like they're all disabled with `fails_with_rgw` :( maybe we could still add a test case and make sure it passes locally, but use `fails_with_rgw` if it's not reliable in teuthology?

cbodley commented on src/test/rgw/rgw_multi/tests.py:
```diff
@@ -2517,6 +2517,86 @@ def check_object_not_exists(bucket, objname):
+    
+    zone_bucket_checkpoint(zoneA, zoneB, bucketA.name)
+    
+    # delete bucket on zoneA. it should fail to delete
+    log.debug('deleting bucket')
+    assert_raises(boto.exception.S3ResponseError, zcA.delete_bucket, bucketA.name)
+
+    assert check_all_buckets_exist(zcA, buckets)
+    assert check_all_buckets_exist(zcB, buckets)
+    return
```
test looks good, but i think it should retry the `delete_bucket` after removing its objects to verify that it succeeds

smanjara replied:
done. b8c9f1416e1def8103c18aa321637013e55b6b55 adds more test cases and accounts for both zonegroup and bucket sync policies.

clwluvw commented:
I happened to come across https://github.com/ceph/ceph/blob/093e0de0b870f3cc10663e24dc60c991ac615612/src/rgw/driver/rados/rgw_rados.cc#L5489, and it seems like it might be redundant. I thought I’d share it here in case it could be considered for cleanup as part of this PR.

cbodley commented on src/rgw/rgw_sync_policy.h:
```diff
@@ -675,6 +675,16 @@ struct rgw_sync_policy_info {
     return groups.empty();
   }
 
+  bool is_directional() const {
+    for (auto& item : groups) {
+      auto& group = item.second;
+      if (!group.data_flow.directional.empty()) {
```
ok, so the presence of a directional flow would mean that we don't have symmetry within the zonegroup. but does `!is_directional()` guarantee symmetry, or is it possible that the group doesn't include all of the zonegroup's zones?

smanjara replied:
yes it is possible that the groups may not contain all zones, which probably means that a zone opted out of sync and it can have its own data. think object prefix filtering also needs to be checked.

smanjara replied:
b3e05c3531ca289f6d90326ee3bf0f6a9fa43adc covers two more cases: zones excluded from sync and object filtering.

smanjara replied:
@cbodley teuthology runs with sync policy test cases haven't been passing consistently. but they pass when I test locally. can you take a look at https://github.com/ceph/ceph/commit/b3e05c3531ca289f6d90326ee3bf0f6a9fa43adc please?

smanjara commented:
@cbodley I think we are good to merge this. what do you feel?

smanjara commented:
https://qa-proxy.ceph.com/teuthology/smanjara-2025-01-23_06:05:25-rgw:multisite-wip-shilpa-cleanup-bi-distro-default-smithi/8090080/teuthology.log

two unrelated tests are failing: `test_copy_object_same_bucket` and` test_copy_object_different_bucket`. need to run them against `main` to see if there is a regression there.

smanjara commented:
> https://qa-proxy.ceph.com/teuthology/smanjara-2025-01-23_06:05:25-rgw:multisite-wip-shilpa-cleanup-bi-distro-default-smithi/8090080/teuthology.log
> 
> two unrelated tests are failing: `test_copy_object_same_bucket` and` test_copy_object_different_bucket`. need to run them against `main` to see if there is a regression there.

directional sync policy configuration was preventing these two tests from passing. looks better after commit 5621d7272f2319c610f4dcfc504e3a82ce25193e :
http://qa-proxy.ceph.com/teuthology/smanjara-2025-02-05_18:45:28-rgw:multisite-wip-shilpa-cleanup-bi-distro-default-smithi/8116198/teuthology.log

smanjara commented:
b1da7d4 and 12b2ffa add sync policy checks along with test cases.  I have covered the following cases where asymmetric replication can occur:
1. zonegroup sync policy directional
2. bucket sync policy directional
3. zones opted out of sync
4. object filter/tags

b1da7d4 covers directional sync at both zonegroup level and bucket level. 
12b2ffa covers zones opted out and also object filtering. if zones have opted out of sync, they won't be part of the sync policy `group` and can have its own data. so, we iterate through all the zones within the zonegroup and check if there is any zone that does not belong to the sync `group`. 
if objects are filtered or tagged for replication, then only those objects sync to other zones. so we query the sync policy to check if object filter/tag is configured on the bucket. 
@soumyakoduri can you please take a look at these two commits.  am I missing any other cases related to sync policy? thanks!

smanjara commented:
> > if objects are filtered or tagged for replication, then only those objects sync to other zones. so we query the sync policy to check if object filter/tag is configured on the bucket.
> 
> my understanding is that these filters only come from bucket replication policy, and i don't think we support that when the source bucket and destination bucket are the same - so within a given bucket, that shouldn't result in different contents on different zones

I see. I could remove that check. but I don't think there is anything preventing a user from 'misconfiguring' the policy that way though.

soumyakoduri commented:
> > > if objects are filtered or tagged for replication, then only those objects sync to other zones. so we query the sync policy to check if object filter/tag is configured on the bucket.
> > 
> > 
> > my understanding is that these filters only come from bucket replication policy, and i don't think we support that when the source bucket and destination bucket are the same - so within a given bucket, that shouldn't result in different contents on different zones
> 
> I see. I could remove that check. but I don't think there is anything preventing a user from 'misconfiguring' the policy that way though.

Right. There is nothing preventing user to create such sync policy (with objects filtered within the same bucket replication). I think even QE validated this using the same source and destination bucket.

smanjara commented:
> > > > if objects are filtered or tagged for replication, then only those objects sync to other zones. so we query the sync policy to check if object filter/tag is configured on the bucket.
> > > 
> > > 
> > > my understanding is that these filters only come from bucket replication policy, and i don't think we support that when the source bucket and destination bucket are the same - so within a given bucket, that shouldn't result in different contents on different zones
> > 
> > 
> > I see. I could remove that check. but I don't think there is anything preventing a user from 'misconfiguring' the policy that way though.
> 
> Right. There is nothing preventing user to create such sync policy (with objects filtered within the same bucket replication). I think even QE validated this using the same source and destination bucket.

if there is no real use-case necessitating it, we should probably avoid filtering across same name buckets at radosgw-admin level. if you both agree, I'll go ahead and make that change.

soumyakoduri commented:
> > > > > if objects are filtered or tagged for replication, then only those objects sync to other zones. so we query the sync policy to check if object filter/tag is configured on the bucket.
> > > > 
> > > > 
> > > > my understanding is that these filters only come from bucket replication policy, and i don't think we support that when the source bucket and destination bucket are the same - so within a given bucket, that shouldn't result in different contents on different zones
> > > 
> > > 
> > > I see. I could remove that check. but I don't think there is anything preventing a user from 'misconfiguring' the policy that way though.
> > 
> > 
> > Right. There is nothing preventing user to create such sync policy (with objects filtered within the same bucket replication). I think even QE validated this using the same source and destination bucket.
> 
> if there is no real use-case necessitating it, we should probably avoid filtering across same name buckets. if you both agree, I'll go ahead and make that change.

I am not sure. TBH, I thought, like in LC, this was intentional to support selective object replication between the same name buckets too. At least that is my understanding from [sync-policy](https://docs.ceph.com/en/latest/radosgw/multisite-sync-policy/#) and  [AWS] (https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication-add-config.html#replication-config-optional-filter)  docs

soumyakoduri commented:
> [b1da7d4](https://github.com/ceph/ceph/commit/b1da7d4a15d12a89ebff0cf7983998554ec062aa) and [12b2ffa](https://github.com/ceph/ceph/commit/12b2ffa86b36ca7180800e3c11301083ef133c63) add sync policy checks along with test cases. I have covered the following cases where asymmetric replication can occur:
> 
>     1. zonegroup sync policy directional
> 
>     2. bucket sync policy directional
> 
>     3. zones opted out of sync
> 
>     4. object filter/tags
> 

Also what about the case of sync from/to different buckets?

smanjara commented:
> > [b1da7d4](https://github.com/ceph/ceph/commit/b1da7d4a15d12a89ebff0cf7983998554ec062aa) and [12b2ffa](https://github.com/ceph/ceph/commit/12b2ffa86b36ca7180800e3c11301083ef133c63) add sync policy checks along with test cases. I have covered the following cases where asymmetric replication can occur:
> > ```
> > 1. zonegroup sync policy directional
> > 
> > 2. bucket sync policy directional
> > 
> > 3. zones opted out of sync
> > 
> > 4. object filter/tags
> > ```
> 
> Also what about the case of sync from/to different buckets?

ok yes, deleting different name source bucket deletes the destination bucket as well. thanks!

smanjara commented:
> > > [b1da7d4](https://github.com/ceph/ceph/commit/b1da7d4a15d12a89ebff0cf7983998554ec062aa) and [12b2ffa](https://github.com/ceph/ceph/commit/12b2ffa86b36ca7180800e3c11301083ef133c63) add sync policy checks along with test cases. I have covered the following cases where asymmetric replication can occur:
> > > ```
> > > 1. zonegroup sync policy directional
> > > 
> > > 2. bucket sync policy directional
> > > 
> > > 3. zones opted out of sync
> > > 
> > > 4. object filter/tags
> > > ```
> > 
> > 
> > Also what about the case of sync from/to different buckets?
> 
> ok yes, deleting different name source bucket deletes the destination bucket as well. thanks!

i'm not sure if this is on purpose or a bug? i don't think we should sync source bucket deletion if it is a different dest bucket.
there could be multiple source buckets syncing to the same destination bucket. cc @likid0

cbodley commented:
what's left to do here? if we're still blocked on the BucketNotEmpty checks for sync policy, i'd suggest that's not worth the extra complexity. i really want to get this fixed for normal replication

smanjara commented:
> what's left to do here? if we're still blocked on the BucketNotEmpty checks for sync policy, i'd suggest that's not worth the extra complexity. i really want to get this fixed for normal replication

done rebasing. w.r.t sync policy checks there is nothing more to be done in this pr.

smanjara commented:
jenkins test make check

smanjara commented on src/test/rgw/rgw_multi/tests.py:
```diff
@@ -2383,6 +2520,28 @@ def remove_sync_policy_group(cluster, group, bucket = None):
+        assert False, 'failed to remove sync flow id=%s, bucket=%s' % (flow, bucket)
+    return json.loads(result_json)
+
+def remove_sync_group_pipe(cluster, group, pipe, bucket = None):
```
resolved

smanjara commented:
> > with [6544c69](https://github.com/ceph/ceph/commit/6544c69e9bf3a8f8af71bb9e33901bfc337fa07b) the tests are passing: https://pulpito.ceph.com/smanjara-2024-12-17_09:39:58-rgw:multisite-wip-shilpa-cleanup-bi-distro-default-smithi/
> 
> i do love to see green multisite runs!
> 
> the test coverage for cleanup on `bilog autotrim` looks good, but we don't have any coverage for the new `BucketNotEmpty` check. i was hoping we could copy one of Soumya's sync policy test cases for this, but it looks like they're all disabled with `fails_with_rgw` :( maybe we could still add a test case and make sure it passes locally, but use `fails_with_rgw` if it's not reliable in teuthology?

i added a few more testcases after this to test sync policy checks. they pass locally. waiting on shaman to build so I can run another round of multisite suite.

smanjara commented:
> > > with [6544c69](https://github.com/ceph/ceph/commit/6544c69e9bf3a8f8af71bb9e33901bfc337fa07b) the tests are passing: https://pulpito.ceph.com/smanjara-2024-12-17_09:39:58-rgw:multisite-wip-shilpa-cleanup-bi-distro-default-smithi/
> > 
> > 
> > i do love to see green multisite runs!
> > the test coverage for cleanup on `bilog autotrim` looks good, but we don't have any coverage for the new `BucketNotEmpty` check. i was hoping we could copy one of Soumya's sync policy test cases for this, but it looks like they're all disabled with `fails_with_rgw` :( maybe we could still add a test case and make sure it passes locally, but use `fails_with_rgw` if it's not reliable in teuthology?
> 
> i added a few more testcases after this to test sync policy checks. they pass locally. waiting on shaman to build so I can run another round of multisite suite.

multisite run looks ok. these are the same failures that are on main. 
https://qa-proxy.ceph.com/teuthology/smanjara-2025-04-03_17:42:14-rgw:multisite-wip-shilpa-cleanup-bi-distro-default-smithi/8223188/teuthology.log
I had to disable sync policy related tests as they are quite flaky on teuthology. passes reliably when tested locally.
